enhancements/proposed/20200501-json-output/start-json-output.md

```diff
@@ -0,0 +1,231 @@
+{"Name":"Done","Message":"🏄  Done! kubectl is now configured to use \"minikube\"","TotalSteps":9,"CurrentStep":9,"Type":"Log"}
+```
+
+This way, clients can parse the output as it is logged and know the following:
+1. What the type of step is (Log vs Download)
+1. What step we are currently on
+1. The total number of steps
+1. The specific message related to that step
+
+
+## Implementation Details
+
+### Stderr
+glog logs can be sent to stderr as usual.
+In addition, we will output error code from [err_map.go](https://github.com/kubernetes/minikube/blob/master/pkg/minikube/problem/err_map.go) as a parsable JSON message to stderr.
+
+This will be done by adding the following function to the `out` package:
+
+```go
+func DisplayErrorJSON(out io.Writer, p *problem.Problem)
+```
+
+which will print the JSON encoding of `p` to `out` (in this case, stderr).
+
+
+### Stdout - Log Steps
+Since we need to approximate the total number of steps before minikube starts, we need to know the general steps we expect to execute before starting.
+
+I propose creating a registry of logs, which is prefilled with the following steps:
+
+* "Minikube Version"
+* "Selecting Driver"
+*	"Starting Control Plane"
+*	"Download Necessary Artifacts"
+*	"Creating Node"
+*	"Preparing Kubernetes"
+*	"Verifying Kubernetes"
+*	"Enabling Addons"
+*	"Done"
+
+When a log is called in the code, it will be associated with one of the above steps.
+
+This will allow us to determine which number step we are currently on, and include that information in the output.
+
+_Note: We may skip steps depending on the user's existing setup. For example, we may not need to "Verify Kubernetes" on a soft start. The current step/total step numbers will only be approximations_
+
+
+This log would be printed at the correct time later in the code by calling a new function, similar to `out.T`:
+
+```go
+func Step(stepName string, style StyleEnum, format string, a ...V)
+```
+
+In the code itself, this means that this line, which currently looks like this:
+
+```go
+out.T(out.Sparkle, `Using the {{.driver}} driver based on existing profile`, out.V{"driver": ds.String()})
+```
+
+would now be:
+
+```go
+out.Step(out.CreateDriver, out.Sparkle, `Using the {{.driver}} driver based on existing profile`, out.V{"driver": ds.String()})
+```
+
+`out.Step` will be responsible for applying the passed in template, and printing out a JSON encoded version of the step if `--output json` is specified:
+
+```json
+{
+  "Name": "Selecting Driver",
+  "Message": "✨  Using the hyperkit driver based on user configuration\n",
+  "TotalSteps": 9,
+  "CurrentStep": 2,
+  "Type": "Log"
+}
+```
+
+### Stdout - Download Steps
+
+To communicate progress on artifacts as they're being downloaded, we want JSON output that looks something like this during download:
+
+```json
+{
+  "Type": "Download",
+  "Artifact": "preload.tar.gz",
```

Comment:
For consumers who don't want to do anything special about download steps, consider keeping a Message field too:
```
"Message": "Downloading preload.tar.gz",
"Artifact": "preload.tar.gz",
```